homeassistant/components/enocean/__init__.py

```diff
@@ -58,23 +58,32 @@ def callback(self, packet):
         is an incoming packet.
         """
         from enocean.protocol.packet import RadioPacket
-        from enocean.utils import combine_hex
         if isinstance(packet, RadioPacket):
             _LOGGER.debug("Received radio packet: %s", packet)
-            for device in self.__devices:
-                if packet.sender_int == combine_hex(device.dev_id):
-                    device.value_changed(packet)
+            self.hass.helpers.dispatcher.dispatcher_send(
+                SIGNAL_RECEIVE_MESSAGE, packet)
 
 
 class EnOceanDevice:
     """Parent class for all devices associated with the EnOcean component."""
 
     def __init__(self, dev_id, dev_name="EnOcean device"):
         """Initialize the device."""
-        ENOCEAN_DONGLE.register_device(self)
         self.dev_id = dev_id
         self.dev_name = dev_name
 
+    async def async_added_to_hass(self):
+        """Register callbacks."""
+        self.hass.helpers.dispatcher.async_dispatcher_connect(
+            SIGNAL_RECEIVE_MESSAGE, self._message_received_callback)
+
+    @callback
+    def _message_received_callback(self, packet):
+        """Handle incoming packets."""
+        from enocean.utils import combine_hex
+        if packet.sender_int == combine_hex(self.dev_id):
+            self.value_changed(packet)
```

Comment:
This function contains home assistant sync api methods. So we shouldn't run `_message_received_callback` in the event loop.

Reply:
I removed the callback